library/src/updater.rs

```diff
@@ -144,14 +145,38 @@ pub fn should_auto_update() -> anyhow::Result<bool> {
     with_config(|config| Ok(config.auto_update))
 }
 
+fn patch_check_request(config: &UpdateConfig, state: &UpdaterState) -> PatchCheckRequest {
+    let latest_patch_number = state.latest_seen_patch_number();
+
+    // Send the request to the server.
+    PatchCheckRequest {
+        app_id: config.app_id.clone(),
+        channel: config.channel.clone(),
+        release_version: config.release_version.clone(),
+        patch_number: latest_patch_number,
+        platform: current_platform().to_string(),
+        arch: current_arch().to_string(),
+    }
+}
+
 fn check_for_update_internal() -> anyhow::Result<PatchCheckResponse> {
-    with_config(|config| {
+    let (request, url, request_fn) = with_config(|config| {
         // Load UpdaterState from disk
         // If there is no state, make an empty state.
         let state =
             UpdaterState::load_or_new_on_error(&config.storage_dir, &config.release_version);
-        send_patch_check_request(config, &state)
-    })
+
+        // Get the required info to make the request.
+        Ok((
+            patch_check_request(config, &state),
+            patches_check_url(&config.base_url),
+            config.network_hooks.patch_check_request_fn,
+        ))
+    })?;
+
+    let response = request_fn(&url, request)?;
+    debug!("Patch check response: {:?}", response);
+    Ok(response)
```

Comment:
How does the response get recorded?  I guess the caller does that?

Reply:
I guess the response never is recorded? TIL.

Reply:
Nope, we don't log the response. What's interesting is that this function is not called by the engine – seemingly only ever by the `shorebird_code_push` package.

Reply:
I think that's why check_for_update_internal exists, was to share code between this (new) function which was exposed for package:shorebird_code_push, and the normal update mechanism.

As I look back through this code I am reminded of how simple/primitive it is.  We just haven't touched it much in a year.

That also might mean that what's happening is that package:shorebird_code_push is running already in a background thread while the user is launching the app (shorebird_init?)

I suspect that we'll also (possibly separately) want to do the fix where we make set_config be able to return quickly without even trying to grab the config lock to help shorebird_init always be fast?  Or maybe we just need to be extra careful to make sure we're never holding the config lock for any length of time (e.g. across a network request).

Reply:
> We just haven't touched it much in a year.

Yep, it's been relatively stable + the consequences of getting it wrong are quite high.

> I suspect that we'll also (possibly separately) want to do the fix where we make set_config be able to return quickly without even trying to grab the config lock to help shorebird_init always be fast

I think we could spawn a new thread when calling `set_config` from `init` to make it look something like:

```rust
pub fn init(
    app_config: AppConfig,
    file_provider: Box<dyn ExternalFileProvider>,
    yaml: &str,
) -> Result<(), UpdateError> {
    #[cfg(any(target_os = "android", test))]
    use crate::android::libapp_path_from_settings;

    init_logging();
    let config = YamlConfig::from_yaml(yaml)
        .map_err(|err| UpdateError::InvalidArgument("yaml".to_string(), err.to_string()))?;

    let libapp_path = libapp_path_from_settings(&app_config.original_libapp_paths)?;
    debug!("libapp_path: {:?}", libapp_path);

    let _ = std::thread::spawn(move || {
        set_config(
            app_config,
            file_provider,
            libapp_path,
            &config,
            NetworkHooks::default(),
        );
    });

    Ok(())
}
```

Reply:
I don't think you want that.  Any time the lock is already taken, there is nothing for us to do.

What about using https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.try_lock in the set_config case?  If the lock is already taken, we know that config is initialized and thus nothing to do (since any other taking of the lock when we're not initialized errors I believe).

Reply:
That would work. We can discuss in the future PR :)